### ISSUE TITLE
Clear up differences between Form and TextField implementation approaches

### ISSUE THREAD

**Issue by matijakevic** [flutter/flutter#133549]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

There are different approaches to changing some specific `StatefulWidget`'s state and it's not clear why is it done that way. At least I cannot find docs/comment about that.

When creating forms, recommended way to access form-related methods is to create `GlobalKey<FormState>` and access them from there.

When using `TextField`, recommended way to change field's value or selection is to use `TextEditingController`.

Question is, how come we don't access `TextField`'s methods using `GlobalKey<TextFieldState>` instead of `TextEditingController`? Or similarly, how come there's no `FormStateController`? Is the reason purely historical?

If writing similar widgets to `Form` and `TextField`, which approach should we prefer?

### Proposal

Add documentation/guidelines on what approach to prefer (and why) when designing more complex widgets like `Form` and `TextField` which require accessing methods on their state to do something related to them.

**Comment by darshankawar** [comment]
@matijakevic Thanks for the report.
Does below documentation for form and textfield help in answering your question / issue ?

https://api.flutter.dev/flutter/widgets/Form-class.html

https://api.flutter.dev/flutter/material/TextField-class.html

https://api.flutter.dev/flutter/widgets/FormState-class.html

**Comment by matijakevic** [comment]
@darshankawar Thank you for the resposne.

Well, not really. The documentation only tells the developer how to use `TextFields` and `Forms`, but now why there are differences in design between the two.

The two approaches are inconsistent. One uses `GlobalKey` to access and modify the state and the other uses _Controller_ approach (which in turn uses notifiers I assume).

Flutter docs never provide any guidelines/docs on which approach to use and why when designing our own widgets which require more complex operations on their state.

**Comment by matijakevic** [comment]
@darshankawar No, unfortunately it doesn't. The docs only provide how to use each of the widgets, not why there are two conceptual differences in the design of those widgets.

**Comment by darshankawar** [comment]
Please check if [this](https://codewithandrea.com/articles/flutter-text-field-form-validation/) article helps further or not.

**Comment by matijakevic** [comment]
No, unfortunately it doesn't.

**Comment by huycozy** [comment]
Perhaps to explain the above questions, need to take a look at design documents or someone from framework team may answer this in more detail. Keeping this as a documentation issue.

**Comment by TahaTesser** [comment]
cc: @LongCatIsLooong

**Comment by LongCatIsLooong** [comment]
One distinction between a TextEditingController and a State associated with a global key is that a TextEditingController allows you to access the editing state of a text field even when the text field isn't present in the widget tree. 
A GlobalKey lets you access a mounted `State` but only if the widget is mounted.

**Comment by matijakevic** [comment]
@LongCatIsLooong Thank you for the response.

Why does this distinction exist? Is there a reason why we shouldn't be able to access form state even when it's unmounted?

**Comment by justinmc** [comment]
I don't know if we have a good place to provide guidance like this, so I think I should close this issue. The best approach will likely depend on your situation.

But if I can give my own thoughts on this, it would be to avoid both approaches if at all possible. Try to avoid any imperative API in your widgets, and instead opt for declaratively passing state into your widgets.
